mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -0,0 +1,217 @@
+      numColBlocks: Int,
+      rdd: RDD[((Int, Int), Matrix)],
+      rowPerBlock: Int,
+      colPerBlock: Int) = {
+    this(numRowBlocks, numColBlocks, rdd)
+    val part = new GridPartitioner(numRowBlocks, numColBlocks, rowPerBlock,
+      colPerBlock, rdd.partitions.length)
+    setPartitioner(part)
+  }
+
+  private[mllib] var partitioner: GridPartitioner = {
+    val firstSubMatrix = rdd.first()._2
+    new GridPartitioner(numRowBlocks, numColBlocks,
+      firstSubMatrix.numRows, firstSubMatrix.numCols, rdd.partitions.length)
+  }
+
+  /**
+   * Set the partitioner for the matrix. For internal use only. Users should use `repartition`.
+   * @param part A partitioner that specifies how SubMatrices are stored in the cluster
+   */
+  private def setPartitioner(part: GridPartitioner): Unit = {
+    partitioner = part
+  }
+
+  private lazy val dims: (Long, Long) = getDim
+
+  override def numRows(): Long = dims._1
+  override def numCols(): Long = dims._2
+
+  /** Returns the dimensions of the matrix. */
+  def getDim: (Long, Long) = {
+    // picks the sizes of the matrix with the maximum indices
+    def pickSizeByGreaterIndex(
+        example: (Int, Int, Int, Int),
+        base: (Int, Int, Int, Int)): (Int, Int, Int, Int) = {
+      if (example._1 > base._1 && example._2 > base._2) {
+        (example._1, example._2, example._3, example._4)
+      } else if (example._1 > base._1) {
+        (example._1, base._2, example._3, base._4)
+      } else if (example._2 > base._2) {
+        (base._1, example._2, base._3, example._4)
+      } else {
+        (base._1, base._2, base._3, base._4)
+      }
+    }
+
+    val lastRowCol = rdd.treeAggregate((0, 0, 0, 0))(
+      seqOp = (c, v) => (c, v) match { case (base, ((blockXInd, blockYInd), mat)) =>
+        pickSizeByGreaterIndex((blockXInd, blockYInd, mat.numRows, mat.numCols), base)
+      },
+      combOp = (c1, c2) => (c1, c2) match {
+        case (res1, res2) =>
+          pickSizeByGreaterIndex(res1, res2)
+      })
+
+    (lastRowCol._1.toLong * partitioner.rowPerBlock + lastRowCol._3,
+      lastRowCol._2.toLong * partitioner.colPerBlock + lastRowCol._4)
+  }
+
+  /** Returns the Frobenius Norm of the matrix */
+  def normFro(): Double = {
+    math.sqrt(rdd.map { mat => mat._2 match {
+      case sparse: SparseMatrix =>
+        sparse.values.map(x => math.pow(x, 2)).sum
+      case dense: DenseMatrix =>
+        dense.values.map(x => math.pow(x, 2)).sum
+    }
+    }.reduce(_ + _))
+  }
+
+  /** Cache the underlying RDD. */
+  def cache(): DistributedMatrix = {
+    rdd.cache()
+    this
+  }
+
+  /** Set the storage level for the underlying RDD. */
+  def persist(storageLevel: StorageLevel): DistributedMatrix = {
+    rdd.persist(storageLevel)
+    this
+  }
+
+  /** Collect the distributed matrix on the driver as a local matrix. */
+  def toLocalMatrix(): Matrix = {
+    val parts = rdd.collect().sortBy(x => (x._1._2, x._1._1))
+    val nRows = numRows().toInt
```

Comment:
I'd check numRows and numCols here before converting to Int and throw an error if the matrix is too large.